src/wx/widgets/sdljoy.cpp

```diff
@@ -38,420 +39,536 @@ static int16_t axisval(int16_t x)
-    }
-
-    if (js_dev) {
-        auto instance = SDL_JoystickInstanceID(js_dev);
 
-        instance_map[instance] = &(joystate[joy]);
+    bool do_poll = false;
+    wxLongLong now = wxGetUTCTimeMillis();
 
-        joystate[joy].instance = instance;
+    if (got_event) {
+        last_poll_ = now;
+    } else if (now - last_poll_ > kPollTimeInterval) {
+        do_poll = true;
+        last_poll_ = now;
     }
 
-    joystate[joy].index = joy;
-    joystate[joy].is_gc = is_gc;
-}
-
-void wxSDLJoy::RemapControllers()
-{
-    for (auto&& joy : joystate) {
-        auto& state = joy.second;
-
-        DisconnectController(state);
-        ConnectController(joy.first);
+    if (do_poll) {
+        for (auto&& joy_state : joystick_states_) {
+            joy_state.second->Poll();
+        }
     }
 }
 
-void wxSDLJoy::DisconnectController(wxSDLJoyState& state)
-{
-    if (auto& dev = state.dev) {
-        if (state.is_gc) {
-            SDL_GameControllerClose(dev);
-        }
-        else {
-            SDL_JoystickClose(dev);
-        }
-
-        dev = nullptr;
-    }
+void wxSDLJoy::ConnectController(int sdl_index) {
+    assert(!add_all_joysticks_);
 
-    instance_map.erase(state.instance);
-}
+    SDL_JoystickID joystick_id;
 
-void wxSDLJoy::Add(int8_t joy_n)
-{
-    if (joy_n < 0) {
-        for (uint8_t joy : range(0, SDL_NumJoysticks()))
-            ConnectController(joy);
+#if SDL_VERSION_ATLEAST(2, 0, 6)
+    joystick_id = SDL_JoystickGetDeviceInstanceID(sdl_index);
+#else
+    // The only way to check the device instance ID prior to SDL 2.0.6 is to
+    // instantiate an SDL_Joystick and get the value.
+    SDL_Joystick* joystick = SDL_JoystickOpen(sdl_index);
+    if (!joystick)
+        return;
+    joystick_id = SDL_JoystickInstanceID(joystick);
+    SDL_JoystickClose(joystick);
+#endif
 
-        add_all = true;
+    if(contains(joystick_states_, joystick_id)) {
+        // Do not reconnect the same device twice. This can happen because
+        // SDL indexes are not stable and change whenever a device is
+        // connected or disconnected.
+        return;
+    }
 
+    auto iter = needed_sdl_indexes_.find(sdl_index);
+    if (iter == needed_sdl_indexes_.end()) {
+        // Only add the joysticks we care about.
         return;
     }
 
-    ConnectController(joy_n);
+    std::unique_ptr<wxSDLJoyState> joy_state(new wxSDLJoyState(sdl_index));
+    if (joy_state->IsValid()) {
+        joystick_states_.emplace(
+            joy_state->joystick_id(), std::move(joy_state));
+        needed_sdl_indexes_.erase(iter);
+    }
 }
 
-void wxSDLJoy::Remove(int8_t joy_n)
-{
-    add_all = false;
-
-    if (joy_n < 0) {
-        for (auto&& joy : joystate)
-            DisconnectController(joy.second);
-
-        joystate.clear();
+void wxSDLJoy::DisconnectController(SDL_JoystickID joystick_index) {
+    assert(!add_all_joysticks_);
 
+    auto iter = joystick_states_.find(joystick_index);
+    if (iter == joystick_states_.end()) {
```

Comment:
Offtopic, but I really don't like this particular idiom in modern C++. Is there something like a convenient optionals header we could use for these operations?

Reply:
Would you prefer

if(map.count(key) == 0)

Reply:
Oh, I think someone already implemented a contains(map, key) function in our code.

Reply:
Yes, I did that, it's a very simple template function.

The issue is doing a lookup that may not exist, and not doing it twice. Languages like Swift use optionals for this.

Reply:
I understand now, I didn't know this was called an "optional".  I was going to say I use this in python a lot, although I'm not sure if those would be considered optionals (like None or inf), but maybe just a natural feature of the duck typing.

Anyway, I found this very recent library https://en.cppreference.com/w/cpp/utility/optional

Reply:
Unfortunately, `std::optional` is C++17.
In this case, however, since we re-use the iterator to delete the element later on, it is faster to keep track of the iterator. I do agree the STL container syntax is less than ideal but it's what we have to deal with.

Reply:
What are the implications for compatibility of switching to C++17? So far I have been keeping it at C++11, but I am clueless about C++ and rarely use advanced features. We do not necessarily have to support 2011 compilers in this year of our lord.

This would only even matter for libretro, and we could keep the core at C++11 or even earlier, and I can make the necessary cmake adjustments for that.

I made this comment on the wrong instance of this usage, I actually meant the the instance id lookups etc.. I realize that in this case you are reusing the iterator and this is all fine.

But in those other cases, I'd like something more elegant. We could switch to C++17 for linux/windows, if everyone is OK with that. If not, I'm sure there is some C++11 compatible header someone wrote for this, perhaps in boost or whatever. If using boost we wouldn't have to depend on distro libraries, we can just bundle the individual headers necessary, there is unlikely to be a runtime dependency. @ZachBacon I'd like your opinion on this too, I know you don't like boost.

I actually consider this important. This is not the only place where we'd need optional syntax, it would be used everywhere in new code. `std::optional` looks very nice actually.

Reply:
I very much am against the use of boost, the core itself needs to remain c++ compatible with older versions due to libretro core being able to be compiled for certain consoles, but the GUI, if we can avoid having another dependency, I very much would appreciate not using boost. That's my personal matter on the situation. It's not that I don't like boost, it's that I just don't want to introduce yet another dependency to vba-m

Reply:
Like I said, if I use boost headers I wouldn't depend on the library.

Reply:
if you think you can get it to work on all platforms the libretro port uses, then go for it. That's the only condition I have.

Reply:
libretro has nothing to do with this, and I said that already.

Reply:
I know, but I am still stating it. If it's just the header, go ahead.